### ISSUE TITLE
BUG: DataFrame.at setter of categorical DF overwrites entire row

### ISSUE THREAD

**Issue by treszkai** [pandas-dev/pandas#37763]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

On a DataFrame with categorical dtype, `df.at[x,y] = v` sets all non-initialized values in row `x`.

#### Code Sample, a copy-pastable example

```
$ python
Python 3.8.0 (default, Oct 28 2019, 16:14:01) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
```

```python
>>> import pandas as pd
>>> pd.__version__
'1.2.0.dev0+1137.g50b34a4a8'
>>> df = pd.DataFrame(index=range(3), columns=range(3), dtype=pd.CategoricalDtype(['foo', 'bar']))
>>> df.at[1,2] = 'foo'
>>> df
     0    1    2
0  NaN  NaN  NaN
1  foo  foo  foo
2  NaN  NaN  NaN
```

It doesn't overwrite values that have been set with `df.loc`:

```python
>>> df = pd.DataFrame(index=range(3), columns=range(3), dtype=pd.CategoricalDtype(['foo', 'bar']))
>>> df.loc[1,1] = 'bar'  # not necessary, just for demo
>>> df.at[1,2] = 'foo'
>>> df
     0    1    2
0  NaN  NaN  NaN
1  foo  bar  foo
2  NaN  NaN  NaN
```

#### Problem description

`df.at[x, y] = v` on a categorical dtype should behave as with other dtypes, and the same as `df.loc[x, y] = v`.

#### Expected Output

The same as what happens with a DF initialized with `None`s:

```python
>>> df = pd.DataFrame([[None] * 3] * 3, index=range(3), columns=range(3), dtype=pd.CategoricalDtype(['foo', 'bar']))
>>> df.loc[1,1] = 'bar'
>>> df.at[1,2] = 'foo'
>>> df
     0    1    2
0  NaN  NaN  NaN
1  NaN  bar  foo
2  NaN  NaN  NaN
```

Or as with `dtype=float`:

```python
>>> df = pd.DataFrame(index=range(3), columns=range(3), dtype=float)
>>> df.loc[1,1] = 1
>>> df.at[1,2] = 27
>>> df
    0    1     2
0 NaN  NaN   NaN
1 NaN  1.0  27.0
2 NaN  NaN   NaN
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
<ul>
<li>------------------</li>
<li>commit           : 50b34a4a8008b6da24c3dac6ae6d66536e463239</li>
<li>python           : 3.8.0.final.0</li>
<li>python-bits      : 64</li>
<li>OS               : Linux</li>
<li>OS-release       : 5.4.0-52-generic</li>
<li>Version          : #57~18.04.1-Ubuntu SMP Thu Oct 15 14:04:49 UTC 2020</li>
<li>machine          : x86_64</li>
<li>processor        : x86_64</li>
<li>byteorder        : little</li>
<li>LC_ALL           : None</li>
<li>LANG             : en_US.UTF-8</li>
<li>LOCALE           : en_US.UTF-8</li>
<li></li>
<li>pandas           : 1.2.0.dev0+1137.g50b34a4a8</li>
<li>numpy            : 1.19.4</li>
<li>pytz             : 2020.4</li>
<li>dateutil         : 2.8.1</li>
<li>pip              : 20.2.4</li>
<li>setuptools       : 50.3.2</li>
<li>Cython           : None</li>
<li>pytest           : None</li>
<li>hypothesis       : None</li>
<li>sphinx           : None</li>
<li>blosc            : None</li>
<li>feather          : None</li>
<li>xlsxwriter       : None</li>
<li>lxml.etree       : None</li>
<li>html5lib         : None</li>
<li>pymysql          : None</li>
<li>psycopg2         : None</li>
<li>jinja2           : None</li>
<li>IPython          : None</li>
<li>pandas_datareader: None</li>
<li>bs4              : None</li>
<li>bottleneck       : None</li>
<li>fsspec           : None</li>
<li>fastparquet      : None</li>
<li>gcsfs            : None</li>
<li>matplotlib       : None</li>
<li>numexpr          : None</li>
<li>odfpy            : None</li>
<li>openpyxl         : None</li>
<li>pandas_gbq       : None</li>
<li>pyarrow          : None</li>
<li>pyxlsb           : None</li>
<li>s3fs             : None</li>
<li>scipy            : None</li>
<li>sqlalchemy       : None</li>
<li>tables           : None</li>
<li>tabulate         : None</li>
<li>xarray           : None</li>
<li>xlrd             : None</li>
<li>xlwt             : None</li>
<li>numba            : None</li>
</ul>
</details>

**Comment by GYHHAHA** [comment]
What's the expected output with `dtype=float`? I think your example returns the right result. @treszkai

**Comment by treszkai** [comment]
Yes, the problem is only with categorical dtype. The Expected Output section compares it to two scenarios that work well, i.e. pre-intialized or non-dtype. Also could've said that "Expected Output: DataFrame.at setter sets a single item, not the entire row."

**Comment by treszkai** [comment]
Setting with `df.at` will call `DataFrame._set_value`, which on the `series = self._get_item_cache(col)` line *Series with the exact same `_values`* regardless if col is 0 or 2:

```python
>>> df = pd.DataFrame(index=range(3), columns=range(3), dtype=pd.CategoricalDtype(['foo', 'bar']))
>>> df.at[0,0] = 'foo'
>>> df[0]._values is df[2]._values
True
```

When setting with `df.loc`, in the `_iLocIndexer._setitem_with_indexer` method `take_split_path` will be true (with categorical dtype), which copies the entire column on every setitem, that's why `loc` isn't affected.
```python
# set the item, possibly having a dtype change
ser = ser.copy()
ser._mgr = ser._mgr.setitem(indexer=pi, value=v)
ser._maybe_update_cacher(clear=True)
```

**Comment by ma3da** [comment]
Yes, I'd say this follows from two things:
* when your DF is instantiated, only one "nan array" is created, which will be associated with all its columns, as they were provided no data ([here](https://github.com/pandas-dev/pandas/blob/c77fc357efbe9443ad8e32b9ce34599e61bcd3f5/pandas/core/internals/construction.py#L271-L272)),
* in the case of categorical columns, the block manager keeps one block per column, which will result in underlying values being the same object (the unique "nan array") ([here](https://github.com/pandas-dev/pandas/blob/c77fc357efbe9443ad8e32b9ce34599e61bcd3f5/pandas/core/internals/managers.py#L1771-L1776)).

I'm tempted to say the way to correct this would be to create one "nan array" per empty column, but there may be a memory-efficient solution.

**Comment by treszkai** [comment]
> to create one "nan array" per empty column, but there may be a
memory-efficient solution.

In what situation is it necessary to spare memory temporarily, until the
elements of the DataFrame are set? When the user asked for a categorical
dtype for the whole DF, it's reasonable to assume that all columns will be
used as such, no?

Luis Pinto <notifications@github.com> schrieb am Mo., 16. Nov. 2020, 10:29:

**Comment by ma3da** [comment]
@treszkai I do agree with you. But I'm not versed enough in the library's internals to feel confident there are no subtleties I'm missing :)

**Comment by ma3da** [comment]
This issue seems to have been solved by #37355.
On master:
```
In [1]: import pandas as pd

In [2]: pd.__version__
Out[2]: '1.2.0.dev0+1320.g28634289c'

In [3]: df = pd.DataFrame(index=range(3), columns=range(3), dtype=pd.CategoricalDtype(['foo', 'bar']))

In [4]: df.at[1,2] = 'foo'

In [5]: df
Out[5]: 
     0    1    2
0  NaN  NaN  NaN
1  NaN  NaN  foo
2  NaN  NaN  NaN
```

**Comment by jreback** [comment]
would take a validation test to close this issue
